proposals/data-classes.md

```diff
@@ -0,0 +1,241 @@
+
+* Database rows
+* JSON/XML messages
+* Login info
+* Configuration options
+
+Data *is not*
+
+* A process
+* A computation
+* A conversation
+* Interactive
+* An object
+
+What you can do with data
+
+* Name it
+* Read it
+* Modify it (or prevent modification)
+* Compose it
+* Compare it
+* Copy it
+
+What you can't do
+
+* Call it
+* Query it
+
+For an object-oriented language this may seem strange, because isn't
+everything an object? In some sense, you can view data as a degenerate object
+-- fields with pure transparency. But this also misses the point. The point
+of object-oriented architecture is to bundle state and behavior and provide
+composable objects that can interactively respond to the system, like cells
+in an organism. There's value in this structure, but it also creates a
+binding between the data and the behavior. By creating data individually we
+allow the data to shift contexts and allow other components to define their
+own behaviors.
+
+## What does it look like in C#?
+
+C# has a couple different ways to represent simple data, but I contend that
+all have fundamental problems.
+
+**Anonymous types**. Things look pretty good at first: you can name, it
+can read it, and you can compare it easily (equality is automatically
+defined). You can't modify it, though -- anonymous types are always immutable
+and you can't easily create a copy with only one change. The real problem
+is composibility. You can't use it as a real type anywhere outside the
+current method and you can't nest it in other data structures, except as
+an object.
+
+**Tuples**. Tuples are a lot like anonymous types. You can provide names,
+read the elements, modify them, and compare them. You can't make them
+immutable, but the real problem is in composition. Tuples aren't really
+an abstraction -- you describe the data structure in full in every
+place you use it. This makes it hard to expand tuples past a certain
+size and makes it difficult to compose with other data structures because
+you cannot refer to them by name.
+
+**Classes/Structs**. This is by far the most common representation of data
+in C#. A canonical example looks something like this:
+
+```C#
+public class LoginResource
+{
+    public string Username { get; set; }
+    public string Password { get; set; }
+    public bool RememberMe { get; set; } = false;
+}
+```
+
+This feature provides names, for both the members and the data structure, it
+provides easy nominal composition, and is easily composible with all other
+data structures. It also provides a convenient syntax for creation by
+interacting directly with the named data, e.g.
+
+```C#
+var x = new LoginResource {
+    Username = "andy",
+    Password = password
+}
+```
+
+Unfortunately, there are still serious problems. There is no piecewise
+comparer implicitly defined for C# classes, so if you want simple data
+comparison, the real example looks like
+[this](https://gist.github.com/agocke/6ba6e64f77f1212ba7292bfd1f1000e5)
```

Comment:
Rather than link to a GIST I would just inline the sample here. Yes it's a big sample but that's part of the point.